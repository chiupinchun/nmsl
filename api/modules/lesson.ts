import { request } from "../core";


export type GetLessonsPayload = {
  tag?: string;
  series?: string;
  search?: string;
  page?: string;
  show?: string;
  order?: string;
};
export interface Comment {
  id: number;
  content: string;
  user: {
    id: string;
    name: string;
  };
  createTime: string;
}
export interface Lesson {
  id: number;
  series: string;
  title: string;
  author: string;
  src: string;
  content: string;
  tags: string;
  goods: number;
  views: number;
  weight: number;
  createTime: string;
  comments: Comment[];
}
export const getLessons = (
  payload: GetLessonsPayload = {}
) => {

  return request<Lesson[]>('/lesson', {
    query: payload
  });
};

export const getPickupLesson = () => getLessons({ show: '5', order: '-weight' });

export const getLessonById = (id: string, cache?: boolean) => {
  return request<Lesson>('/lesson/' + id, { cache: cache ? 'force-cache' : 'no-cache' });
};

export const getComments = (
  payload: {
    page?: string;
    show?: string;
    lessonId: number;
  }
) => {
  const { lessonId, ...others } = payload;
  return request<Comment[]>('/lesson-comment/' + lessonId, { query: others });
};