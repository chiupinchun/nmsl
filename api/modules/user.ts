import { request } from "../core";

export interface LoginPayload {
  account: string;
  password: string;
}
export const login = (payload: LoginPayload) => {
  return request('/user/login', {
    method: 'POST',
    body: payload
  });
};

export interface SignupPayload extends LoginPayload {
  name: string;
}
export const signup = (payload: SignupPayload) => {
  return request<{
    account: string;
    createTime: string;
    id: string;
    name: string;
  }>('/user', {
    method: 'POST',
    body: payload
  });
};