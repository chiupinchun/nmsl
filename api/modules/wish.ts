import { request } from "../core";

export const postWish = (wish: string) => {
  return request('/wish', {
    method: 'POST',
    body: { wish }
  });
};

export interface Wish {
  wish: string;
  count: number;
}
export const getWishes = () => {
  return request<Wish[]>('/wish');
};