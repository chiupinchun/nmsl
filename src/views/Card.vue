<template>
  <div class="background">
    <div class="container">
      <Transition>
        <Suspense>
          <component :is="editMode ? UserInfoForm : UserInfoCard" :data="userData" @generate="generateCard"
            @back="editMode = true"></component>
        </Suspense>
      </Transition>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref, reactive } from 'vue';
import UserInfoForm from '../components/card/UserInfoForm.vue';
import UserInfoCard from '../components/card/UserInfoCard.vue';

const editMode = ref(true);
const userData = reactive<UserInfo | {}>({});

const generateCard = ($event: UserInfo) => {
  editMode.value = false;
  Object.assign(userData, $event);
};
</script>

<style lang="less" scoped>
.background {
  display: flex;
  justify-content: center;
  align-items: center;
  // flex-direction: column;
  width: 100vw;
  height: 100vh;
  background: url(@/assets/images/bg.jpg);

  .container {
    padding: 2.5rem;
    width: 50vw;
    border: 1px solid black;
    border-radius: 5px;
    background-color: rgb(220, 220, 255);
  }
}

.v-enter-active,
.v-leave-active,
.v-enter-from,
.v-leave-to {}

.v-enter-from,
.v-leave-to {
  transition: 0.5s;
  opacity: 0;
  position: absolute;
}
</style>