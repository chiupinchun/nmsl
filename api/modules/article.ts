import { request } from "../core";
import { UserInfo } from "./user";

export const typeOptions = ['技術', '發問', '徵才'] as const;
export const techOptions = ['前端', '後端', '全端', '爬蟲', 'AI', '區塊鏈', '算法', '運維', '網安', '測試'] as const;

export const postArticle = (payload: {
  title: string;
  type: string;
  content: string;
  tech?: string;
}) => {
  return request('/article', {
    method: 'POST',
    body: payload
  });
};

export interface Article {
  id: number;
  type: string;
  title: string;
  content: string;
  tech: string;
  user: Partial<UserInfo>;
  createTime: string;
}
export const getArticles = (payload: {
  type?: typeof typeOptions[number];
  tech?: typeof techOptions[number];
  show?: string;
} = {}) => {
  return request<Article[]>('/article', {
    query: payload
  });
};

export interface Comment {
  id: number;
  content: string;
  user: {
    id: string;
    name: string;
  };
  createTime: string;
}
interface ArticleWithComment extends Article {
  comments: Comment[];
}
export const getArticleById = (id: string, cache?: boolean) => {
  return request<ArticleWithComment>('/article/' + id, { next: { revalidate: cache ? 60 * 60 : 0 } });
};
