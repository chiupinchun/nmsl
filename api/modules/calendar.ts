import { request } from "../core";
import { UserInfo } from "./user";

export interface Calendar {
  id: number;
  user: UserInfo;
  title: string;
  content: string;
  date: string;
  createTime: string;
}
export const getCalendars = () => {
  return request<Calendar[]>('/calendar');
};

export interface AddCalendarPayload {
  title: string;
  content: string;
  date: Date;
}
export const addCalendar = (payload: AddCalendarPayload) => {
  return request('/calendar', {
    method: 'POST',
    body: payload
  });
};

export interface EditCalendarPayload extends AddCalendarPayload {
  id: number;
}
export const editCalendar = (payload: EditCalendarPayload) => {
  const { id, ...body } = payload;

  return request(`/calendar/${id}`, {
    method: 'PATCH',
    body: body
  });
};

export const delCalendar = (id: number) => {
  return request(`/calendar/${id}`, {
    method: 'DELETE'
  });
};