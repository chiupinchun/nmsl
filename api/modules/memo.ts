import { request } from "../core";
import { UserInfo } from "./user";

export interface Memo {
  id: number;
  user: UserInfo;
  title: string;
  content: string;
  type: 'todo' | 'doing' | 'done';
  createTime: string;
}
export const getMemos = () => {
  return request<Memo[]>('/memo');
};

interface AddMemoPayload {
  title: string;
  content: string;
  type: Memo['type'];
}
export const addMemo = (payload: AddMemoPayload) => {
  return request('/memo', {
    method: 'POST',
    body: payload
  });
};

interface EditMemoPayload extends AddMemoPayload {
  id: number;
}
export const editMemo = (payload: EditMemoPayload) => {
  const { id, ...body } = payload;

  return request(`/memo/${id}`, {
    method: 'PATCH',
    body: body
  });
};

export const delMemo = (id: number) => {
  return request(`/memo/${id}`, {
    method: 'DELETE'
  });
};