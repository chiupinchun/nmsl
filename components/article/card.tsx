"use client";
import { useState, type FC } from 'react';
import { Card, CardContent, CardDescription, CardFooter, CardHeader, CardTitle } from '../ui/card';
import Link from 'next/link';
import { type Article, deleteArticle as _deleteArticle } from '@/api/modules/article';
import { Badge } from '../ui/badge';
import { cn, marked } from '@/lib/utils';
import {
  Popover,
  PopoverContent,
  PopoverTrigger,
} from "@/components/ui/popover";
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from "@/components/ui/dialog";
import { Pencil } from 'lucide-react';
import { useSelector } from '@/store';
import { Button } from '../ui/button';
import { useToast } from '../ui/use-toast';
import { useRouter } from 'next/navigation';

interface Props {
  article: Article;
  className?: string;
  inList?: boolean;
  changeMode?: () => void;
}

const page: FC<Props> = ({ article, className, inList, changeMode }) => {
  const router = useRouter();
  const user = useSelector(state => state.user);
  const { toast } = useToast();

  const [showDeleteDialog, setShowDeleteDialog] = useState(false);
  const deleteArticle = async () => {
    const res = await _deleteArticle(article.id);
    if (res?.success) {
      toast({ description: '刪除成功' });
      router.push('/article/list');
    } else toast({ variant: 'destructive', description: res?.message ?? '發生錯誤！' });
  };

  return (
    <>
      <Card className={cn('relative w-screen', className)}>
        <CardHeader>
          <div className='flex justify-between items-center'>
            <div className='flex items-center'>
              <CardTitle><Link href={`/article/${article.id}`}>【{article.type}】{article.title}</Link></CardTitle>
              {
                user?.id === article.user.id && changeMode && (
                  <Popover>
                    <PopoverTrigger>
                      <Pencil className='ms-2 cursor-pointer' size='20' />
                    </PopoverTrigger>
                    <PopoverContent className='w-[unset]'>
                      <Button onClick={changeMode} className='me-3'>編輯</Button>
                      <Button onClick={() => setShowDeleteDialog(true)} variant='destructive'>刪除</Button>
                    </PopoverContent>
                  </Popover>
                )
              }
            </div>
            {article.tech && <Badge className='cursor-pointer'><Link href={`/article/list?tech=${article.tech}`}>{article.tech}</Link></Badge>}
          </div>
          <CardDescription className='flex justify-between'>
            <Link href={`/resume/${article.user.id}`}>{article.user.name}</Link>
            <span>{new Date(article.createTime).toLocaleDateString()}</span>
          </CardDescription>
        </CardHeader>
        <CardContent dangerouslySetInnerHTML={{ __html: marked(article.content) }} className={cn(inList ? 'max-h-72 overflow-hidden' : '', 'markdown-body')}></CardContent>
        {inList && <CardFooter className='absolute bottom-0 right-2'>
          <Link href={`/article/${article.id}`} className='p-2 bg-slate-900 rounded-xl text-slate-300 font-bold text-sm transition-all hover:text-base'>完整內容&gt;&gt;</Link>
        </CardFooter>}
      </Card>
      <Dialog open={showDeleteDialog} onOpenChange={show => setShowDeleteDialog(show)}>
        <DialogContent>
          <DialogHeader>確定要刪除文章嗎？</DialogHeader>
          <DialogDescription>刪除後將不可復原，請謹慎選擇。</DialogDescription>
          <div className='flex space-x-5'>
            <Button onClick={deleteArticle} variant='destructive'>刪除</Button>
            <Button onClick={() => setShowDeleteDialog(false)}>取消</Button>
          </div>
        </DialogContent>
      </Dialog>
    </>
  );
};

export default page;