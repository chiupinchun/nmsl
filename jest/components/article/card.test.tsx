import '@testing-library/jest-dom';
import { render, screen, within } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import Card from '@/components/article/card';
import { type Article } from '@/api/modules/article';

describe('card component', () => {
  const mockArticle: Article = {
    id: 1,
    type: '技術',
    title: '測試',
    content: '測試',
    tech: '測試',
    user: {
      name: 'Hato',
      checkable: true
    },
    createTime: new Date().toLocaleDateString()
  };

  it('could render', () => {
    render(<Card article={mockArticle} />);
    expect(Card).toMatchSnapshot();
  });

  it('could display as list item', () => {
    render(<Card article={mockArticle} inList={true} />);
    expect(screen.getByRole('link', { name: /完整內容/ })).toBeInTheDocument();
  });

  it('could display as single card', () => {
    render(<Card article={mockArticle} />);
    expect(screen.queryByRole('link', { name: /完整內容/ })).not.toBeInTheDocument();
  });

  it('should display edit btn only when login', async () => {
    render(<Card article={mockArticle} />);

    expect(screen.queryByRole('button', { name: '編輯' })).not.toBeInTheDocument();
  });

  it('could toggle edit mode', async () => {
    const toggle = jest.fn();
    mockArticle.user.id = 'bbb28023-73ba-4e47-842b-b33cca1ee2fa';

    render(<Card article={mockArticle} changeMode={toggle} />);

    const title = screen.getByRole('heading', { name: new RegExp(mockArticle.title) });
    const popoverBtn = within(title.parentElement!).getByRole('button');
    await userEvent.click(popoverBtn);

    await userEvent.click(screen.getByRole('button', { name: '編輯' }));

    expect(toggle).toHaveBeenCalled();
  });
});