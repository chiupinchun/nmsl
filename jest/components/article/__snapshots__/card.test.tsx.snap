// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card component could render 1`] = `[Function]`;
